contributors/design-proposals/containerized-mounter-pod.md

```diff
@@ -0,0 +1,200 @@
+```
+mount: wrong fs type, bad option, bad superblock on 192.168.0.1:/test_vol,
+       missing codepage or helper program, or other error
+```
+
+Kubelet will periodically retry mounting the volume and it will eventually succeed when pod with mount utilities is scheduled and running on the node.
+
+### gRPC API
+
+`VolumeExecService` is a simple gRPC service that allows to execute anything via gRPC:
+
+```protobuf
+service VolumeExecService {
+    // Exec executes a command and returns its output.
+    rpc Exec(ExecRequest) returns (ExecResponse) {}
+}
+
+message ExecRequest {
+    // Command to execute
+    string cmd = 1;
+    // Command arguments
+    repeated string args = 2;
+}
+
+message ExecResponse {
+    enum ExecError {
+        // Helps to reconstruct exec.ErrNotFound.
+        COMMAND_NOT_FOUND = 1;
+        // Helps to reconstruct exec.ExitError.
+        EXIT_CODE = 2;
+    }
+    Error error = 1;
+    // Exit code of the command. This field is nozero only when Error == EXIT_CODE.
+    int32 exit_code = 2;
+    // Capture of combined stdout + stderr.
+    string output = 3;
+}
+```
+
+* Both `ExecRequest` and `ExecResponse` are tailored for execution of mount utilities that don't need any stdin and stdout+stderr are typically short. Therefore there is no streaming of these file descriptors.
+
+* No authentication / authorization is done on the server side, anyone who connects to the socket can execute anything. It is expected that only root has access to `/var/lib/kubelet/plugin-sockets/`.
+
+* `.proto` file for this API will be stored in `k8s.io/kubernetes/pkg/version/apis/exec/v1alpha1`.
+
+  * `hack/update-generated-runtime.sh` will be updated to generate go files for this API.
+
+    * Should it be renamed to `update-generated-grpc-apis.sh`?
+
+* Kubernetes will ship a daemon with server implementation of this API in `cmd/volume-exec`. This implementation simply calls `os.Exec` for each `ExecRequest` it gets and returns the right response.
+
+  * Authors of container images with mount utilities can then add this `volume-exec` daemon to their image, they don't need to care about anything else.
+
+### Upgrade
+Upgrade of the DaemonSet with pods with mount utilities needs to be done node by node and with extra care. The pods may run fuse daemons and killing such pod with glusterfs fuse daemon would kill all pods that use glusterfs on the same node.
+
+In order to update the DaemonSet, admin must do for every node:
+* Mark the node as tainted. Only the pod with mount utilities can tolerate the taint, all other pods are evicted. As result, all volumes are unmounted and detached.
+* Update the pod.
+* Remove the taint.
+
+Is there a way how to make it with DaemonSet rolling update? Is there any other way how to do this upgrade better?
```

Comment:
I am asking if there are some tricks how to do DaemonSet rolling update that would drain a node first before updating the pod. Otherwise I need to fall back to `--cascade=false` and do the update manually as @humblec suggests.